GVFS/FastFetch/NativeUnixMethods.cs

```diff
@@ -0,0 +1,146 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.ComponentModel;
+using System.IO.MemoryMappedFiles;
+using System.Runtime.InteropServices;
+
+namespace FastFetch
+{
+    public class NativeUnixMethods 
+    {
+        public const int ReadOnly = 0x0000;
+        public const int WriteOnly = 0x0001;
+
+        public const int Create = 0x0200;
+        public const int Truncate = 0x0400;
+        private const int InvalidFileDescriptor = -1;
+
+        public static unsafe void WriteFile(ITracer tracer, byte* originalData, long originalSize, string destination, ushort mode)
+        {
+            int fileDescriptor = InvalidFileDescriptor;
+            try
+            {
+                fileDescriptor = Open(destination, WriteOnly | Create | Truncate, mode);
+                if (fileDescriptor != InvalidFileDescriptor)
+                {
+                    IntPtr result = Write(fileDescriptor, originalData, (IntPtr)originalSize);
+                    if (result.ToInt32() == -1)
+                    {
+                        throw new Win32Exception(Marshal.GetLastWin32Error());
+                    }
+                }
+            }
+            catch (Win32Exception e)
+            {
+                tracer.RelatedError("Error writing file {0}. Win32Exception: {1}", destination, e);
+                throw;
+            }
+            finally
+            {
+                Close(fileDescriptor);
+            }
+        }
+
+        public static bool StatAndUpdateIndexForFile(string path, MemoryMappedViewAccessor indexView, long offset)
+        {
+            NativeStat.StatBuffer st = StatFile(path);

```

Comment:
Should we be catching exceptions thrown by `StatFile` and returning false? 

Related: I noticed the Windows implementation returns false if the file does not exist and we're not doing that here.

Reply:
Updated it to catch the exception and take a tracer. We'll get an exception from stat if the file doesn't exist so we should be covered for that case. Alternatively, we can test to see if the file exists here, but this is a very hot path so I'm a little concerned about perf.